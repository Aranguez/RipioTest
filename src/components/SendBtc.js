import React, { Component } from 'react';

import Resumen from './Resumen';
import MontoTotal from './MontoTotal';
import Loading from './Loading';
import Formulario from './Formulario';

export default class SendBtc extends Component {

    state = {
        totalAmount: 2000,
        address: '',
        amount: '',
        fee: parseFloat((Math.random() / 1000).toFixed(5)),
        errorAddress: {
            error: false,
            msg: ''
        },
        errorAmount: {
            error: false,
            msg: ''
        },
        envios: []
    }

    checkForm = (addressBtc, amount) => { //checkea errores del formulario
        if (addressBtc.length === 0) {
            this.setState({
                errorAddress: {
                    error: true,
                    msg: 'ingrese un adrress btc'
                }
            })
            return false
        } else if (addressBtc !== 'address correcto'){
            this.setState({
                errorAddress: {
                    error: true,
                    msg: 'address incorrecto'
                }
            })
            return false
        } else if (amount > this.state.totalAmount) {
            console.error('no puede exceder su monto')
            this.setState({
                errorAmount: {
                    error: true,
                    msg: 'no puede exceder su monto'
                }
            })
            return false
        } else if (amount === 0 || this.state.amount === ''){
            console.error('ingrese un monto')
            this.setState({
                errorAmount: {
                    error: true,
                    msg: 'ingrese un monto'
                }
            })
            return false
        } else {
            return true
        }
    }

    sendBitcoin = data => { //envia bitcoin
        this.setState({
            loading: true
        })

        setTimeout(() => {
          this.setState( prevState => ({
            address: '',
            amount: '',
            errorAddress: {
                error: false,
                msg: ''
            },
            errorAmount: {
                error: false,
                msg: ''
            },
            envios: [...prevState.envios, data],
            totalAmount: (prevState.totalAmount - data.amount).toFixed(5),
            loading: false
          }))
        }, 1500)
    }
    
    handleOnSubmit = e => { // trae los datos del form, los checkea y envia
        e.preventDefault();
        let addressBtc = this.state.address;
        let amount = parseFloat(this.state.amount);

        if (this.checkForm(addressBtc, amount)) {
            let data = {
                addressBtc,
                amount: amount + this.state.fee,
                totalAmount: this.state.totalAmount - (amount + this.state.fee)
            }
            this.sendBitcoin(data)
        } else {
            return
        }
    }

    handleOnChange = e => { //maneja los values con el state
        this.setState({
            [e.target.name]: e.target.value
        })
    }

    render() {  
        
        console.log(this.state)
        return (
            <div className="container">
                <div className="row">
                    <div className="col-xs-12">
                        <MontoTotal totalAmount={this.state.totalAmount}/>
                    </div>
                </div>
                <div className="row">
                    <div className="col-xs-12 col-sm-6">
                        { !this.state.loading ?
                            <Formulario
                                handleOnSubmit={this.handleOnSubmit}
                                handleOnChange={this.handleOnChange}
                                address={this.state.address}
                                amount={this.state.amount}
                                errorAmount={this.state.errorAmount}
                                errorAddress={this.state.errorAddress}
                                fee={this.state.fee}
                                /> :
                            <Loading/> }
                    </div>
                    <div className="col-xs-12 col-sm-6">
                        { this.state.envios.length === 0 ?
                            <div className="box">
                                <h4 className="text-center">No hay envios recientes</h4>
                            </div> :
                            
                            this.state.envios.map( (envio, index) => {
                                return (
                                    <Resumen
                                        key={index}
                                        address={envio.addressBtc}
                                        amount={envio.amount}
                                        totalAmount={envio.totalAmount}/>
                                    )
                            })
                        }
                    </div>

                    
                </div>
            </div>
        )
    }
}
