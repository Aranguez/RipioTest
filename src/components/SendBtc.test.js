import React from 'react';
import { shallow, mount } from 'enzyme';
import SendBtc from './SendBtc';
import { spy } from 'sinon';

describe('SendBtc', () => {
    
    const sendBtc = shallow(<SendBtc/>);

    it('renderiza sin problemas', () => {
        expect(sendBtc).toMatchSnapshot();
    })

    const wallet = 2000;
    const fee = 0.00012;
    const amount = 22;

    const data = {
        addressBtc: 'address correcto',
        amount: amount,
        totalAmount: wallet - (amount + fee)
    }

    it('inicializa el `state` con una lista de envios vacia', () => {
        expect(sendBtc.state().envios).toEqual([])
    })

    //test onHandleOnChange function
    describe('cuando el usuario tipea en los inputs', () => {
        //mock de la funcion
        sendBtc.instance().handleOnChange = jest.fn();
        sendBtc.update();

        it('modifica el address en el `state`', () => {
            const event = { target: {
                value: data.addressBtc,
                name: 'address'
            }};
            sendBtc.find('input[name="address"]').simulate('change', event);
            expect(sendBtc.state().address).toEqual(event.target.value);
        })
    })

    // test checkForm function
    /*describe('cuando ejecuta `checkForm()`', () => {
        //mock de la funcion
        sendBtc.instance().checkForm = jest.fn();
        sendBtc.instance().handleOnSubmit = jest.fn();
        sendBtc.update();

        it('llama a la funcion', () => {

            const event = {
                preventDefault() {},
            };

            sendBtc.find('form').simulate('submit', event)
            expect(sendBtc.instance().checkForm)
                .toBeCalledWith(sendBtc.state().address, sendBtc.state().amount)
        })

        //check errores en diferentes casos
    })*/

    //test sendBitcoin function
    /*describe('cuando envia Bitcoin', () => {
        sendBtc.instance().sendBitcoin = jest.fn();
        sendBtc.update();

        it('llama a la funcion con la data', () => {
            expect(sendBtc.instance().sendBitcoin).toBeCalledWith(data)
        })

        it('agrega un envio al state `state`', () => {
            expect(sendBtc.state().envios).toEqual([{data}])
        })
    })*/

    //test del formulario
    describe('cuando clickea en enviar', () => {

        beforeAll(() => {
            sendBtc.setState({
                address: data.addressBtc,
                amount: data.amount,
                fee: fee
            })
        })

        it('check addressBtc', () => {
            expect(sendBtc.state().address).toEqual('address correcto');
        })

        it('check if amount it is less or equal than wallet', () => {
            expect(data.amount).toBeLessThanOrEqual(wallet)
        })

        it('check if amount is equal to 0', () => {
            expect(data.amount).not.toBe(0)
        })

        it('submit event when click submit', () => {
            sendBtc.instance().onHandleSubmit = jest.fn();
            sendBtc.update();

            const submit = sendBtc.instance().onHandleSubmit;
            const wrapper = mount(
                <form onSubmit={submit} id="form">
                    <div className="form-group">
                        <label>Address BTC</label>
                        <input
                            className="input"
                            placeholder="address btc"
                            name="address"
                            defaultValue={data.addressBtc} 
                            />
                    </div>
                    
                    <div className="form-group">
                        <label>Amount to send</label>
                        <input
                            className="input"
                            placeholder="amount to send"
                            name="amount"
                            defaultValue={data.amount}
                            />
                        <small>Fee: {fee}</small>
                    </div>
                    
                    <button
                        className="button btn-azul submit"
                        type="submit">Enviar</button>
                </form>
            );
            wrapper.find('[type="submit"]').simulate('submit');
            expect(submit).toBeCalled()
        });

    })
})

