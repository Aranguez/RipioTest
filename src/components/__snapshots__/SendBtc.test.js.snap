// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendBtc renderiza sin problemas 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SendBtc />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <MontoTotal
          totalAmount={2000}
        />,
        <form
          id="form"
          onSubmit={[Function]}
        >
          <div
            className="form-group"
          >
            <label>
              Address BTC
            </label>
            <input
              className="input"
              name="address"
              onChange={[Function]}
              placeholder="address btc"
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <label>
              Amount to send
            </label>
            <input
              className="input"
              name="amount"
              onChange={[Function]}
              placeholder="amount to send"
              value={0}
            />
            <small>
              Fee: 
              0.00032
            </small>
          </div>
          <button
            className="button btn-azul submit"
            type="submit"
          >
            Enviar
          </button>
        </form>,
        Array [],
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "totalAmount": 2000,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="form-group"
            >
              <label>
                Address BTC
              </label>
              <input
                className="input"
                name="address"
                onChange={[Function]}
                placeholder="address btc"
                value=""
              />
            </div>,
            <div
              className="form-group"
            >
              <label>
                Amount to send
              </label>
              <input
                className="input"
                name="amount"
                onChange={[Function]}
                placeholder="amount to send"
                value={0}
              />
              <small>
                Fee: 
                0.00032
              </small>
            </div>,
            <button
              className="button btn-azul submit"
              type="submit"
            >
              Enviar
            </button>,
          ],
          "id": "form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label>
                  Address BTC
                </label>,
                <input
                  className="input"
                  name="address"
                  onChange={[Function]}
                  placeholder="address btc"
                  value=""
                />,
              ],
              "className": "form-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Address BTC",
                },
                "ref": null,
                "rendered": "Address BTC",
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "input",
                  "name": "address",
                  "onChange": [Function],
                  "placeholder": "address btc",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label>
                  Amount to send
                </label>,
                <input
                  className="input"
                  name="amount"
                  onChange={[Function]}
                  placeholder="amount to send"
                  value={0}
                />,
                <small>
                  Fee: 
                  0.00032
                </small>,
              ],
              "className": "form-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Amount to send",
                },
                "ref": null,
                "rendered": "Amount to send",
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "input",
                  "name": "amount",
                  "onChange": [Function],
                  "placeholder": "amount to send",
                  "value": 0,
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Fee: ",
                    0.00032,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "Fee: ",
                  0.00032,
                ],
                "type": "small",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Enviar",
              "className": "button btn-azul submit",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Enviar",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MontoTotal
            totalAmount={2000}
          />,
          <form
            id="form"
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <label>
                Address BTC
              </label>
              <input
                className="input"
                name="address"
                onChange={[Function]}
                placeholder="address btc"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label>
                Amount to send
              </label>
              <input
                className="input"
                name="amount"
                onChange={[Function]}
                placeholder="amount to send"
                value={0}
              />
              <small>
                Fee: 
                0.00032
              </small>
            </div>
            <button
              className="button btn-azul submit"
              type="submit"
            >
              Enviar
            </button>
          </form>,
          Array [],
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "totalAmount": 2000,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="form-group"
              >
                <label>
                  Address BTC
                </label>
                <input
                  className="input"
                  name="address"
                  onChange={[Function]}
                  placeholder="address btc"
                  value=""
                />
              </div>,
              <div
                className="form-group"
              >
                <label>
                  Amount to send
                </label>
                <input
                  className="input"
                  name="amount"
                  onChange={[Function]}
                  placeholder="amount to send"
                  value={0}
                />
                <small>
                  Fee: 
                  0.00032
                </small>
              </div>,
              <button
                className="button btn-azul submit"
                type="submit"
              >
                Enviar
              </button>,
            ],
            "id": "form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label>
                    Address BTC
                  </label>,
                  <input
                    className="input"
                    name="address"
                    onChange={[Function]}
                    placeholder="address btc"
                    value=""
                  />,
                ],
                "className": "form-group",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Address BTC",
                  },
                  "ref": null,
                  "rendered": "Address BTC",
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "input",
                    "name": "address",
                    "onChange": [Function],
                    "placeholder": "address btc",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label>
                    Amount to send
                  </label>,
                  <input
                    className="input"
                    name="amount"
                    onChange={[Function]}
                    placeholder="amount to send"
                    value={0}
                  />,
                  <small>
                    Fee: 
                    0.00032
                  </small>,
                ],
                "className": "form-group",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Amount to send",
                  },
                  "ref": null,
                  "rendered": "Amount to send",
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "input",
                    "name": "amount",
                    "onChange": [Function],
                    "placeholder": "amount to send",
                    "value": 0,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Fee: ",
                      0.00032,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    "Fee: ",
                    0.00032,
                  ],
                  "type": "small",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Enviar",
                "className": "button btn-azul submit",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Enviar",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
